drivers/pwm/pwm_stm32_lptim.c

```diff
@@ -0,0 +1,249 @@
+				   ? LL_LPTIM_OUTPUT_POLARITY_INVERSE
+				   : LL_LPTIM_OUTPUT_POLARITY_REGULAR;
+		/* set to max to have 100% duty cycle (of inverse) */
+		pulse_cycles = 0xffff;
+	} else if (pulse_cycles == period_cycles) {
+		/* set to max to have 100% duty cycle */
+		pulse_cycles = 0xffff;
+	} else {
+		/* remove 1 pulse cycle, accounts for 1 extra low cycle */
+		pulse_cycles -= 1U;
+	}
+
+	/* remove 1 period cycle, accounts for 1 extra low cycle */
+	period_cycles -= 1U;
+
+	LL_LPTIM_OC_SetPolarity(timer, ll_channel, polarity);
+	set_timer_compare[channel - 1u](timer, pulse_cycles);
+	LL_LPTIM_SetAutoReload(timer, period_cycles);
+
+	if (!LL_LPTIM_CC_IsEnabledChannel(timer, ll_channel)) {
+		LL_LPTIM_CC_SetChannelMode(timer, ll_channel, LL_LPTIM_CCMODE_OUTPUT_PWM);
+		LL_LPTIM_CC_EnableChannel(timer, ll_channel);
+	}
+
+	return 0;
+}
+
+static int pwm_stm32_lptim_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					      uint64_t *cycles)
+{
+	struct pwm_stm32_lptim_data *data = dev->data;
+	const struct pwm_stm32_lptim_config *cfg = dev->config;
+
+	*cycles = (uint64_t)(data->tim_clk / cfg->prescaler);
+
+	return 0;
+}
+
+static DEVICE_API(pwm, pwm_stm32_lptim_driver_api) = {
+	.set_cycles = pwm_stm32_lptim_set_cycles,
+	.get_cycles_per_sec = pwm_stm32_lptim_get_cycles_per_sec,
+};
+
+static int pwm_stm32_lptim_init(const struct device *dev)
+{
+	struct pwm_stm32_lptim_data *data = dev->data;
+	const struct pwm_stm32_lptim_config *cfg = dev->config;
+	LPTIM_TypeDef *timer = cfg->timer;
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	uint32_t tim_clk;
+	int r;
+
+	if (!device_is_ready(clk)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	/* Enable clock and store its speed */
+	r = clock_control_on(clk, (clock_control_subsys_t)&cfg->pclken[0]);
+	if (r < 0) {
+		LOG_ERR("Could not initialize clock (%d)", r);
+		return r;
+	}
+
+	if (cfg->pclk_len > 1) {
+		/* Enable Timer clock source */
+		r = clock_control_configure(clk, (clock_control_subsys_t)&cfg->pclken[1], NULL);
+		if (r != 0) {
+			LOG_ERR("Could not configure clock (%d)", r);
+			return r;
+		}
+
+		r = clock_control_get_rate(clk, (clock_control_subsys_t)&cfg->pclken[1], &tim_clk);
+		if (r < 0) {
+			LOG_ERR("Timer clock rate get error (%d)", r);
+			return r;
+		}
+	} else {
+		LOG_ERR("Timer clock source is not specified");
+		return -EINVAL;
+	}
+
+	data->tim_clk = tim_clk;
+
+	/* Reset timer to default state using RCC */
+	(void)reset_line_toggle_dt(&data->reset);
+
+	/* Configure pinmux */
+	r = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (r < 0) {
+		LOG_ERR("PWM pinctrl setup failed (%d)", r);
+		return r;
+	}
+
+	LOG_INF("Clock rate: %u Hz, Prescaler: %u", data->tim_clk, cfg->prescaler);
+
+	/* Initialize timer */
+	LL_LPTIM_SetClockSource(timer, LL_LPTIM_CLK_SOURCE_INTERNAL);
+	LL_LPTIM_SetPrescaler(timer, (__CLZ(__RBIT(cfg->prescaler)) << LPTIM_CFGR_PRESC_Pos));
```

Comment:
Nitpicking: could remove an inner parentheses pair:
```suggestion
	LL_LPTIM_SetPrescaler(timer, __CLZ(__RBIT(cfg->prescaler)) << LPTIM_CFGR_PRESC_Pos);